### ISSUE TITLE
Using take() and repeat() with share() results in the entire stream being repeated.

### ISSUE THREAD

**Issue by thepont** [ReactiveX/rxjs#2886]
**RxJS version:** 5.4.3

**Code to reproduce:**
```js
Observable
    .from([1,2,3])
    .share()
    .take(1)
    .repeat(10)
    .do((ii) => console.log(ii))
    .toPromise()
```

**Expected behavior:**
```
1
1
1
1
1
1
1
1
1
1
```
**Actual behavior:**
```
1
2
3
1
2
3
1
2
```
**Additional information:**

The way the repeat function appears to work also makes it hard to create Observables using `Observable.create` since the function passed to create has to be certain to make certain that the cursor for whatever it's putting in the stream is reset each time. The rxjs-stream project suffers from this when it attempts to convert nodejs streams into Observables, this would also be true for rx-node if it was converted to RXv5.

For instance.

This would work.
```js
function TestObservable(){
    return Rx.Observable.create((observer) => {
        var it = makeIterator(['hello', 'world', 'what', 'is', 'happening']);
        let current = it.next()
        if(current.done){
            observer.complete();
        } else {
            observer.next(current.value);
        }
    })
}
```
While this wouldn't
```js
function TestObservable(){
    var it = makeIterator(['hello', 'world', 'what', 'is', 'happening']);
    return Rx.Observable.create((observer) => {
        let current = it.next()
        if(current.done){
            observer.complete();
        } else {
            observer.next(current.value);
        }
    })
}
```
And nor would
```js
function TestObservable(){
    return Rx.Observable.create((observer) => {
        var it = makeIterator(['hello', 'world', 'what', 'is', 'happening']);
        let current = it.next()
        if(current.done){
            observer.complete();
        } else {
            observer.next(current.value);
        }
    }).share()
}
```
Where the test function is:

```js
TestObservable().take(1).repeat(10).do((ii) => console.log(ii)).toPromise()
```

and the `makeIterator` function is.

```js
function makeIterator(array) {
    var nextIndex = 0;
    return {
       next: function() {
           return nextIndex < array.length ?
               {value: array[nextIndex++], done: false} :
               {done: true};
       }
    };
}
```

**Comment by benlesh** [comment]
You're sharing a synchronous observable. `Observable.from(arrayHere)` will iterate over that array with a `for` loop on subscription. That means that by the time the call to `source.subscribe()` is over, it's already done emitting all values, and is unsubscribed.

`share` uses refCounting under the hood. So the refCount goes to 1, then synchronously goes down to 0.  When the refCount hits 0 in share, the underlying Subject is recycled, to allow repeat subscriptions. This means the observable goes "cold" again, waiting for a new "first subscriber".

... `repeat` detects the complete and basically concatenates the source (shared observable) onto the end again. The source just recycled the subject and will resubscribe to the original observable.

If you use, say, an `interval`, you'll see what you're expecting. Because that's async.


TL;DR: `share` does effectively nothing to totally synchronous observables.

If you want to mutlicast something synchronously, I recommend using the `publish(selector)` operator.
